index/scorch/builder.go

```diff
@@ -0,0 +1,303 @@
+
+	return o.maybeFlushBatchLOCKED()
+}
+
+func (o *Builder) maybeFlushBatchLOCKED() error {
+	if len(o.batch.IndexOps) >= o.batchSize {
+		defer o.batch.Reset()
+		return o.executeBatchLOCKED(o.batch)
+	}
+	return nil
+}
+
+func (o *Builder) executeBatchLOCKED(batch *index.Batch) (err error) {
+	analysisResults := make([]*index.AnalysisResult, 0, len(batch.IndexOps))
+	for _, doc := range batch.IndexOps {
+		if doc != nil {
+			// insert _id field
+			doc.AddField(document.NewTextFieldCustom("_id", nil, []byte(doc.ID), document.IndexField|document.StoreField, nil))
+			// perform analysis directly
+			analysisResult := analyze(doc)
+			analysisResults = append(analysisResults, analysisResult)
+		}
+	}
+
+	seg, _, err := zap.AnalysisResultsToSegmentBase(analysisResults, DefaultChunkFactor)
+	if err != nil {
+		return fmt.Errorf("error building segment base: %v", err)
+	}
+
+	filename := zapFileName(o.segCount)
+	o.segCount++
+	path := o.buildPath + string(os.PathSeparator) + filename
+	err = zap.PersistSegmentBase(seg, path)
+	if err != nil {
+		return fmt.Errorf("error persisting segment base to %s: %v", path, err)
+	}
+
+	o.segPaths = append(o.segPaths, path)
+
+	return nil
+}
+
+func (o *Builder) doMerge() error {
+	// as long as we have more than 1 segment, keep merging
+	for len(o.segPaths) > 1 {
+
+		// merge the next <mergeMax> number of segments into one new one
+		// or, if there are fewer than <mergeMax> remaining, merge them all
+		mergeCount := o.mergeMax
+		if mergeCount > len(o.segPaths) {
+			mergeCount = len(o.segPaths)
+		}
+
+		mergePaths := o.segPaths[0:mergeCount]
+		o.segPaths = o.segPaths[mergeCount:]
+
+		// open each of the segments to be merged
+		mergeSegs := make([]*zap.Segment, 0, mergeCount)
+		closeOpenedSegs := func() {
+			for _, seg := range mergeSegs {
+				_ = seg.Close()
+			}
+		}
+		for _, mergePath := range mergePaths {
+			seg, err := zap.Open(mergePath)
+			if err != nil {
+				closeOpenedSegs()
+				return fmt.Errorf("error opening segment (%s) for merge: %v", mergePath, err)
+			}
+			mergeSegs = append(mergeSegs, seg.(*zap.Segment))
+		}
+
+		// do the merge
+		mergedSegPath := o.buildPath + string(os.PathSeparator) + zapFileName(o.segCount)
+		drops := make([]*roaring.Bitmap, mergeCount)
+		_, _, err := zap.Merge(mergeSegs, drops, mergedSegPath, DefaultChunkFactor, nil, nil)
+		if err != nil {
+			closeOpenedSegs()
+			return fmt.Errorf("error merging segments (%v): %v", mergePaths, err)
+		}
+		o.segCount++
+		o.segPaths = append(o.segPaths, mergedSegPath)
+
+		// close segments  opened for merge
+		closeOpenedSegs()
+
+		// remove merged segments
+		for _, mergePath := range mergePaths {
+			err = os.RemoveAll(mergePath)
+			if err != nil {
+				return fmt.Errorf("error removing segment %s after merge: %v", mergePath, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (o *Builder) Close() error {
+	o.m.Lock()
+	defer o.m.Unlock()
+
+	// see if there is a partial batch
+	err := o.maybeFlushBatchLOCKED()
```

Comment:
@mschoch maybe I'm reviewing to quickly and I'm missing something, but it looks to me like if I call `Close()` when there is some documents in the batch (ex. index 999 documents, with default 1000 batch, those would not get flushed ?)

Reply:
yep good catch

Reply:
added a test for this, and adjusted the logic